### ISSUE TITLE
Type load exception on Dilithium with Dilithium config disabled

### ISSUE THREAD

**Issue by WizX20** [SitecoreUnicorn/Unicorn#249]
Sitecore 8.2.170728
Unicorn 4.0.2: Unicorn.Dilithium.config has been disabled.

`[TypeLoadException: Declaration referenced in a method implementation cannot be a final method.  Type: 'Unicorn.Data.Dilithium.DilithiumSitecoreDataStore'.  Assembly: 'Unicorn, Version=4.0.2.0, Culture=neutral, PublicKeyToken=null'.]
   Unicorn.Data.Dilithium.DilithiumExtensions.EnablesDilithiumSql(IConfiguration configuration) +0
   Unicorn.ControlPanel.Controls.ConfigurationInfo.Render(HtmlTextWriter writer) +833
   Unicorn.ControlPanel.Responses.ControlPanelPageResponse.Execute(HttpResponseBase response) +379
   Unicorn.ControlPanel.UnicornControlPanelPipelineProcessor.ProcessRequest(HttpContext context) +690
   Unicorn.ControlPanel.UnicornControlPanelPipelineProcessor.Process(HttpRequestArgs args) +94`

(Sorry for the dutch in the message there)

Just upgraded the Unicorn packages, am I missing something?

**Comment by kamsar** [comment]
It looks like an assembly might be out of date? Check the deployed version.
Either way some config is still referencing the dilithium data store; it is
possible to enable it without the config file, which just changes the
defaults to dilithium

**Comment by WizX20** [comment]
I re-fetched the sources, removed obj and cleared the bin directories. Clean and rebuild, removed all binaries and configs related with rainbow and unicorn. And finally re-deployed. It does not seem to help. The deployed version is correct.

Weird, I installed it on a vanilla Sitecore of the same version and it seems to work. Well, back to the drawing board!

**Comment by WizX20** [comment]
Ok, I've played around with it a bit and it now crashes on the Sitecore vanilla installation too.

The additional files:
[App_Config-Include.zip](https://github.com/kamsar/Unicorn/files/1352011/App_Config-Include.zip)

- \Include
  - Rainbow.config: default rainbow configuration
  - Unicorn.Helix.config: Copied from Habitat source
- \Include\Foundation
  - Foundation.Serialization.config: Foundation set-up as defined in our solution
  - Foundation.Serialization.Settings.config: max path lenght and truncation settings
- \Include\Project
  - Website.Tenant.config: The tenant config, only contains a sitecore var for the solutions sourceFolder (Make sure this is an existing path on disk, and for the foundation serialization add the path $(sourcefolder)\Foundation\Serialization\serialization on disk)
- \Include\Unicorn
  - Default configuration with an additional Unicorn.zSharedSecret.config secret key patch.

**Comment by WizX20** [comment]
Uncommented al the include nodes in Foundation.Serialization.config, then there is a logical exception that there are no nodes :)

Enabled as only include:
`<include name="Settings.Feature" database="master" path="/sitecore/system/Settings/Feature">
				<exclude children="true" />
			</include>
			<include name="Settings.Foundation" database="master" path="/sitecore/system/Settings/Foundation">
				<exclude children="true" />
			</include>
			<include name="Settings.Project" database="master" path="/sitecore/system/Settings/Project">
				<exclude children="true" />
			</include>`

Still no-go, this also breaks Unicorn.

**Comment by kamsar** [comment]
In the original exception, what would the message be in English? Google Translate nets me the lovely and I'm sure incorrect `TypeLoadException: The declaration referred to in a method implementation must not be a termination method`

**Comment by WizX20** [comment]
Try to get Windows to use English instead of the default installed language..anyway I worked it out this is the message in English:

`Declaration referenced in a method implementation cannot be a final method.  Type: 'Unicorn.Data.Dilithium.DilithiumSitecoreDataStore'.  Assembly: 'Unicorn, Version=4.0.2.0, Culture=neutral, PublicKeyToken=null'.`

**Comment by WizX20** [comment]
Found it! NuGet failed to update the reference to the new Kamsar.WebConsole.dll. In visual studio the reference was ok, but when I cleared out my packages folder I noticed the reference went into broken state... so when upgrading re-check your references to the rainbow and unicorn binaries.

**Comment by kamsar** [comment]
How odd, well glad you figured it out!

**Comment by cassidydotdk** [comment]
Reopening this to investigate. Just did a 3.3.2 upgrade to 4.0.3 and was hit by this error. 


Declaration referenced in a method implementation cannot be a final method.  Type: 'Unicorn.Data.Dilithium.DilithiumSitecoreDataStore'.  Assembly: 'Unicorn, Version=4.0.3.0, Culture=neutral, PublicKeyToken=null'.
Description: An unhandled exception occurred during the execution of the current web request. Please review the stack trace for more information about the error and where it originated in the code. 

Exception Details: System.TypeLoadException: Declaration referenced in a method implementation cannot be a final method.  Type: 'Unicorn.Data.Dilithium.DilithiumSitecoreDataStore'.  Assembly: 'Unicorn, Version=4.0.3.0, Culture=neutral, PublicKeyToken=null'.

Source Error: 

An unhandled exception was generated during the execution of the current web request. Information regarding the origin and location of the exception can be identified using the exception stack trace below.

Stack Trace: 


[TypeLoadException: Declaration referenced in a method implementation cannot be a final method.  Type: 'Unicorn.Data.Dilithium.DilithiumSitecoreDataStore'.  Assembly: 'Unicorn, Version=4.0.3.0, Culture=neutral, PublicKeyToken=null'.]
   Unicorn.Data.Dilithium.DilithiumExtensions.EnablesDilithiumSql(IConfiguration configuration) +0
   Unicorn.ControlPanel.Controls.ConfigurationInfo.Render(HtmlTextWriter writer) +772
   Unicorn.ControlPanel.Responses.ControlPanelPageResponse.Execute(HttpResponseBase response) +379
   Unicorn.ControlPanel.UnicornControlPanelPipelineProcessor.ProcessRequest(HttpContext context) +710
   Unicorn.ControlPanel.UnicornControlPanelPipelineProcessor.Process(HttpRequestArgs args) +94
   (Object , Object[] ) +74
   Sitecore.Pipelines.CorePipeline.Run(PipelineArgs args) +469
   Sitecore.Pipelines.DefaultCorePipelineManager.Run(String pipelineName, PipelineArgs args, String pipelineDomain) +22
   Sitecore.Nexus.Web.HttpModule.(Object , EventArgs ) +517
   System.Web.SyncEventExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute() +139
   System.Web.HttpApplication.ExecuteStepImpl(IExecutionStep step) +195
   System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously) +88

**Comment by cassidydotdk** [comment]
Definitely some assembly conflict nonsense. It caught me out though. I copied debug builds of all Unicorn/Rainbow assemblies to the target bin folder and the error went away. I then deleted my debug assemblies, redeployed the solution that was failing like above - and the error is still gone.

I did do a Kamsaw.Webconsole updated manually as described above, but that did not make the problem go away.

Maybe Configy?
